Trace.py

```diff
@@ -55,18 +55,28 @@ def _dump_pfb(xbox):
     return bytes(buffer)
 
 
-def _read_pgraph_rdi(xbox: Xbox, offset: int, count: int) -> bytes:
+def _read_pgraph_rdi(xbox: Xbox, offset: int, count: int):
     # FIXME: Assert pusher access is disabled
     # FIXME: Assert PGRAPH idle
 
     NV10_PGRAPH_RDI_INDEX = 0xFD400750
     NV10_PGRAPH_RDI_DATA = 0xFD400754
 
-    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
-
-    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
-
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, original_offset)
+    # JFR: Reading the DATA register 4 times returns X,Y,Z,W (not in that order as far
+    # as I remember), but during that time the INDEX register will stay constant, only
+    # on the final read I think it auto-increments the INDEX.
+    # It is not safe and likely incorrect to do a bulk read so this must be done
+    # individualy despite the interface communication overhead.
```

Comment:
I can remove the attribution if you'd prefer (it does mean that there's a chance someone will come ask me via git blame and I'd send them to look at this PR if I didn't already know), but I am both reluctant to make this PR any more complicated than it already is (assuming after testing I do find some way to speed things up) and concerned about not leaving some sort of explanation behind. 

I think without a comment here there's a high chance that another contributor will follow in my footsteps; see that RDI dumping is so slow that unpredictable instability prevents capture of even a single frame, see that there's a tight loop instead of a bulk read, and erroneously attempt a bulk read to remove overhead.

As context, both myself and @antangelo have experienced transient failures that seem unpredictable and manifest as either socket timeouts or (more often) getmem failures in `xboxpy`. Running the same capture in as close as possible to the same state often leads to different results though, making this very hard to diagnose.
As further context, dumping a single frame of Morrowind (even with my broken RDI approach which sped up the read to the point of insignificance) took over three hours in the attempt that worked, with at least another hour on attempts that failed along the way.

My worry is that, without some warning, there will be a very strong drive to improve performance by folks who are way less knowledgeable about the internals than the maintainers and it'll be quite discouraging to contribution if they put in significant effort (e.g., if they don't already know Python or aren't programmers by trade) and learn after the fact that their change isn't actually viable.

Reply:
> If you want to keep it similar, then please remove my name & rephrase it.

Totally understand, attribution has been removed and I did some rewording.

> However, I'd suggest to remove the comment entirely, or look up for a trustable source like nouveau, or run your own experiments and document how hardware behaves (potentially on XboxDevWiki or nouveau / envytools, and then link the corresponding docs from nv2a-trace).

This is the intent of the `#TODO`; once I cycle back to this I'll dig deeper to verify the behavior (I'd be surprised if it turns out to be something other than your previous experience).

> I remember traces taking long for me, but not that long. I usually connect my Xbox directly to my laptop for nv2a-trace though. However, chances are that I was also running a different xboxpy, seeing how many branches of it I have locally. I also remember messing a lot with xbdm and other networking protocols to speed it up. There's at least one commit in [XboxDev/xboxpy#25](https://github.com/XboxDev/xboxpy/pull/25) but I don't know if it helps. There might be more elsehwere.

Cool, I'll look at that next time I'm doing a dump. I'm also testing with a hard wired connection so I doubt network latency attributable to the physical connection has any meaningful impact.

> Unfortunately I've lost track of most xboxpy / nv2a-trace stuff over the years, because nobody else was trying it, so I stopped bothering with pushing or documenting my findings. I digress again, but when I was working on some GPU related demos I would also run the nxdk code on the PC using a modified pbkit / XGU and inject the generated pushbuffers via network. So I'd run the logic on PC, and inject pushbuffers into different emulators and real consoles by broadcasting it via network, then reading the framebuffer remotely to verify emulation / drivers. Most of my Xbox work in the end was automated and done remotely without touching the console. My console was headless most of the time. I might clean up and push these things sometime, but I probably won't have time / interest anytime soon.

This is a fascinating approach and there are some parallel discussions around how to set up a combination of null graphics emulators and GitHub-action driven hardware to do CI on Discord. If you do find the time I think there may be a number of newer contributors to various projects (like myself) who would be extremely interested in catching up on the results of your previous efforts!